akka-stream/src/main/scala/akka/stream/scaladsl/One2OneBidiFlow.scala

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream._
+
+import scala.util.control.NoStackTrace
+
+/**
+ * A generic BidiFlow which verifies that another flow produces exactly one output element per
```

Comment:
I guess the implementation problem is that the bypass operates asynchronously, so you don't know in which order you will receive input and output elements. I wonder if there's an ever simpler implementation which puts a simple side-effect stage before the wrapped flow that increments an AtomicInteger, and a similar stage after the wrapped flow that decrements the same counter and then
 * throws if the counter falls below 0
 * asserts that the counter is zero at completion

Shouldn't this work as well? Not sure, how much you need to have to work around materialization to share the counter between the two stages (in the worst case the component becomes non-reusable).